sdk/core/core-https/src/pipeline.ts

```diff
@@ -0,0 +1,285 @@
+   * the request is performed.
+   */
+  getOrderedPolicies(): PipelinePolicy[];
+  /**
+   * Duplicates this pipeline to allow for modifying an existing one without mutating it.
+   */
+  clone(): Pipeline;
+}
+
+interface PipelineDescriptor {
+  policy: PipelinePolicy;
+  options: AddPolicyOptions;
+}
+
+interface PolicyGraphNode {
+  policy: PipelinePolicy;
+  dependsOn: Set<PolicyGraphNode>;
+  dependants: Set<PolicyGraphNode>;
+  afterPhase?: Set<PolicyGraphNode>;
+}
+
+/**
+ * A private implementation of Pipeline.
+ * Do not export this class from the package.
+ * @internal
+ */
+export class HttpsPipeline implements Pipeline {
+  private _policies: PipelineDescriptor[] = [];
+  private _orderedPolicies?: PipelinePolicy[];
+
+  private constructor(policies: PipelineDescriptor[] = []) {
+    this._policies = policies;
+    this._orderedPolicies = undefined;
+  }
+
+  public addPolicy(policy: PipelinePolicy, options: AddPolicyOptions = {}): void {
+    if (options.phase && options.afterPhase) {
+      throw new Error("Policies inside a phase cannot specify afterPhase.");
+    }
+    if (options.phase && !ValidPhaseNames.has(options.phase)) {
+      throw new Error(`Invalid phase name: ${options.phase}`);
+    }
+    if (options.afterPhase && !ValidPhaseNames.has(options.afterPhase)) {
+      throw new Error(`Invalid phase name: ${options.afterPhase}`);
+    }
+    this._policies.push({
+      policy,
+      options
+    });
+    this._orderedPolicies = undefined;
+  }
+
+  public removePolicy(options: { name?: string; phase?: string }): PipelinePolicy[] {
+    const removedPolicies: PipelinePolicy[] = [];
+
+    this._policies = this._policies.filter((policyDescriptor) => {
+      if (
+        (options.name && policyDescriptor.policy.name === options.name) ||
+        (options.phase && policyDescriptor.options.phase === options.phase)
+      ) {
+        removedPolicies.push(policyDescriptor.policy);
+        return false;
+      } else {
+        return true;
+      }
+    });
+    this._orderedPolicies = undefined;
+
+    return removedPolicies;
+  }
+
+  public sendRequest(
+    httpsClient: HttpsClient,
+    request: PipelineRequest
+  ): Promise<PipelineResponse> {
+    const policies = this.getOrderedPolicies();
+
+    const pipeline = policies.reduceRight<SendRequest>(
+      (next, policy) => {
+        return (request: PipelineRequest) => {
+          return policy.sendRequest(request, next);
+        };
+      },
+      (request: PipelineRequest) => httpsClient.sendRequest(request)
+    );
+
+    return pipeline(request);
+  }
+
+  public getOrderedPolicies(): PipelinePolicy[] {
+    if (!this._orderedPolicies) {
+      this._orderedPolicies = this.orderPolicies();
+    }
+    return this._orderedPolicies;
+  }
+
+  public clone(): Pipeline {
+    return new HttpsPipeline(this._policies);
+  }
+
+  public static create(): Pipeline {
+    return new HttpsPipeline();
+  }
+
+  private orderPolicies(): PipelinePolicy[] {
```

Comment:
Even though this function is well structured and readable, I think that having some comments through what's going on would make it easier to follow.

Reply:
yes, let me attempt to document it better in comments!